neural_speed/application/main_pybind.cpp

```diff
@@ -54,6 +56,241 @@
+  }
+
+  sequence Query2Sequence(const Query& query) {
+    sequence ret_seq;
+    ret_seq.request_idx = -1;  // let scheduler decides it
+    ret_seq.prompt_ids = query.token_ids;
+    ret_seq.n_prompt_tokens = query.token_ids.size();
+    ret_seq.n_tokens = query.token_ids.size();
+    ret_seq.n_past = 0;
+    ret_seq.n_total = 0;
+    ret_seq.gen_conf.max_new_tokens = params.n_predict;
+    ret_seq.gen_conf.min_new_tokens = params.min_new_tokens;
+    ret_seq.gen_conf.length_penalty = params.length_penalty;
+    ret_seq.gen_conf.do_early_stopping = params.do_early_stopping;
+    ret_seq.query_id = query.id;
+    return ret_seq;
+  }
+
+  Query Sequence2Query(const sequence& seq) {
+    Query ret_query;
+    ret_query.id = seq.query_id;
+    int ret_size = return_prompt ? seq.prompt_ids.size() + seq.generated_ids.size() : seq.generated_ids.size();
+    ret_query.token_ids.resize(ret_size);
+    if (return_prompt) {
+      std::copy(seq.prompt_ids.cbegin(), seq.prompt_ids.cend(), ret_query.token_ids.begin());
+      std::copy(seq.generated_ids.cbegin(), seq.generated_ids.cend(),
+                ret_query.token_ids.begin() + seq.prompt_ids.size());
+    } else {
+      std::copy(seq.generated_ids.cbegin(), seq.generated_ids.cend(), ret_query.token_ids.begin());
+    }
+    return ret_query;
+  }
+
+  void InitServerParams(const std::string& model_path, int max_new_tokens, int n_batch, int ctx_size, int seed,
+                        int threads, float repetition_penalty, int num_beams, bool do_sample, int top_k, float top_p,
+                        float temperature, int min_new_tokens, float length_penalty, bool early_stopping, int n_keep,
+                        int n_discard, bool shift_roped_k, int batch_size, model_vocab::id pad_token,
+                        const std::string& memory_dtype, const bool& continuous_batching, const int& max_request_num,
+                        const float& model_scratch_enlarge_scale) {
+    init_gpt_params(&params, model_path, max_new_tokens, n_batch, ctx_size, seed, threads, repetition_penalty,
+                    num_beams, do_sample, top_k, top_p, temperature, min_new_tokens, length_penalty, early_stopping,
+                    n_keep, n_discard, shift_roped_k, batch_size, pad_token, memory_dtype, continuous_batching,
+                    max_request_num, model_scratch_enlarge_scale);
+  }
+
+  ~ModelServer() {
+    // "synchronized" function
+    // stop spinning after calling ResponseCallback for the last query
+    py::print("Stopping CPP server...");
+    running = false;
+    {
+      py::gil_scoped_release releaser;
+      worker.join();
+    }
+    py::print("CPP server stopped!");
+  }
+
+ private:
+  const ResponseCallback response;
+  std::vector<Query> waiting;
+  std::mutex queue_mtx;
+  bool running;
+  gpt_params params;
+  std::string policy;
+  bool scheduler_empty;
+  uint64_t working_size;
+  bool return_prompt;
+  std::thread worker;
```

Comment:
some comments here might be helpful

Reply:
will do.

Reply:
added in https://github.com/intel/neural-speed/pull/69/commits/949a463ef3412dcae2971989b678ffd3f62d3271